servers/physics_3d/godot_space_3d.cpp

```diff
@@ -840,12 +844,30 @@ bool GodotSpace3D::test_body_motion(GodotBody3D *p_body, const PhysicsServer3D::
 				}
 
 				int shape_idx = intersection_query_subindex_results[i];
+				Transform3D col_obj_xform = col_obj->get_transform() * col_obj->get_shape_transform(shape_idx);
 
+#ifndef USE_ITERATIVE
+				real_t max_advance;
+				bool collided = GodotCollisionSolver3D::sweep(body_shape, body_shape_xform, col_obj->get_shape(shape_idx), col_obj_xform, p_parameters.motion, max_advance, motion_aabb, 0.001);
+
+				if (collided) {
+					if (max_advance == 0.0) {
+						stuck = true;
+						break;
+					}
+
+					real_t fraction = p_parameters.motion.length() / 64.0;
```

Comment:
I think the new safe/unsafe logic here is the source of the problem. The `fraction` variable is being used for linear offsets to the `safe` and `unsafe` proportions, and so `fraction` should also be a proportion; however it's actually just 1/64th of the total sweep length. If the sweep length is >= 64 then fraction becomes 1>= and the safe/unsafe values get pegged to 0 and 1.

Reply:
It is not strictly related with the sweep test itself,  but in most modern sliding algorithms is its considered good practice to add an extra offset (Or extra fraction) so the body does not get pushed all the way towards the surface. This can be done in the movement algorithm as an extra step, but would require recalculating this fractions and adjusting the the fractions and movement, but it would be adding complexity to the final piece.

Move and collide already has the concept of the safe margin. The reason it doesn't fulfill our necessities here is that it works pre movement, so it will still get pushed by the geometry in there is any movement towards it before being corrected (Not always on the right direction) and cause and extra iteration on the sliding algorithm that would otherwise not be needed.

I think it would make sense to consider  implementing this into move and collide for an easier use of it. This would make move and collide more stable and reliable at the time of keeping this safe distance against the geometry. Here is an extremely simple character made with move and collide. A safe margin is being used to prevent the body from getting too close and create problems with the movement. As you can see, it results in stutter and unreliable behavior (The character floats 0.5 meters over the ground and it only moves on the horizontal plane with move and collide, there is no gravity while on floor).

https://github.com/godotengine/godot/assets/64601424/88f2dc0d-4c7e-408c-a775-13dc00db3903

As you can see, even in the most simple case, the fact that we are using the safe margin only before movement, but is not being taking into account to actually prevent the movement from violating it, results in this kind of problems.